pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java

```diff
@@ -65,6 +67,7 @@ public ImmutableSegmentImpl(SegmentDirectory segmentDirectory, SegmentMetadataIm
     _segmentMetadata = segmentMetadata;
     _indexContainerMap = columnIndexContainerMap;
     _starTreeIndexContainer = starTreeIndexContainer;
+    _lazyDataSource = new HashMap<>(segmentMetadata.getColumnMetadataMap().size());
```

Comment:
We can create all the data sources up-front in the constructor instead of creating them lazily. If we create them lazily, we need to use a concurrent map here, which might impact performance; also, the first query hitting the column could be slower than other queries.

Reply:
You are right. I didn't think about concurrency.